### PR TITLE
Input from EM4100 tag reader

### LINKED ISSUE #17: Make it work with custom RFID reader

### PR DESCRIPTION
**Pull request by emanuelen5**
Closes #17

### CODE REVIEW

File: src/util/key_reader.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+    @classmethod
+    def get_reader(cls):
+        key_readers = cls.get_devices()
+        if len(key_readers) == 0:
+            raise NoReaderFound()
+
+        key_reader = key_readers[0]
+        if len(key_readers) > 1:
+            logger.warning(f"There are several key readers connected: {key_readers}. Choose {key_reader}")
```
Comment (emanuelen5) [inline src/util/key_reader.py]: Misspelling (Choose=Chose)

File: src/util/key_reader.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    def __repr__(self):
+        return f"<EM4100 Key Reader tty={self.serial_device}>"
+
+    def tag_was_read(self):
+        if self.com.in_waiting == 0:
+            return False
+        lines = self.com.read(10000).decode("utf-8").split("\r\n")
+        complete_readouts = []
+        for line in lines:
+            match = re.match(r"^DECODED: MANCHESTER=(0x[a-fA-F0-9]{8})$", line)
```
Comment (emanuelen5) [inline src/util/key_reader.py]: Only Manchester decoding is checked here. We should also check for Biphase

File: src/util/key_reader.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        self.check_echo()
+        self.last_tag_id = None
+
+    def __repr__(self):
+        return f"<EM4100 Key Reader tty={self.serial_device}>"
+
+    def tag_was_read(self):
+        if self.com.in_waiting == 0:
+            return False
+        lines = self.com.read(10000).decode("utf-8").split("\r\n")
```
Comment (emanuelen5) [inline src/util/key_reader.py]: 10000 is arbitrarily chosen. We can only get 115200 byte per second at most from the serial port.

File: src/util/key_reader.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+    @classmethod
+    def get_devices(cls):
+        output = subprocess.check_output(LIST_ARDUINO_SERIAL_DEVICES_PATH).decode("utf-8")
+        devices = [dev for dev in output.split("\n") if len(dev) > 0]
+        key_readers = []
+        for dev in devices:
+            try:
+                key_readers.append(cls(dev))
+            except Exception as e:
```
Comment (emanuelen5) [inline src/util/key_reader.py]: I don’t know what type pf exception to expect here. This was just a precaution.

File: src/util/key_reader.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+class NoReaderFound(Exception):
+    pass
+
+def abstract(fun):
+    def abstract_wrapper(*args, **kwargs):
+        raise NotImplementedError("This is only and abstract function")
+    return abstract_wrapper
+
+class KeyReader(object):
```
Comment (emanuelen5) [inline src/util/key_reader.py]: My thought is that we might have several implementations of key readers (Aptus, HID devices, EM4100 etc.), which is what this interface class should abstract.

File: src/util/key_reader.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
+from config import LIST_ARDUINO_SERIAL_DEVICES_PATH
+from src.util.logger import get_logger
+import subprocess
+import serial
+import re
+
+logger = get_logger()
+
+class NoReaderFound(Exception):
```
Comment (emanuelen5) [inline src/util/key_reader.py]: Should perhaps exchange the inherited error type to something more descriptive.

File: memberbooth.py [hunk 7]
```diff
@@ -40,7 +41,13 @@ def main():
         logger.error("The makeradmin client is not logged in")
         sys.exit(-1)
 
-    app = Application(makeradmin_client)
+    try:
+        key_reader = EM4100_KeyReader.get_reader()
+    except NoReaderFound as e:
```
Comment (InMyOrbit) [inline memberbooth.py]: If --development is used, I think that the application should run even if there is no reader attached.

File: src/util/key_reader.py (hunk 5 above)
Comment (InMyOrbit) [reply src/util/key_reader.py]: Sounds good.

File: memberbooth.py (hunk 7 above)
Comment (InMyOrbit) [reply memberbooth.py]: The function below also has to be changed to do nothing on timer expired, or maybe even simpler, never start the timer if the --development flag is specified. 


https://github.com/makerspace/memberbooth/pull/35/files#diff-17273fda14cf37080e22b106abbe439eR99

Comment (emanuelen5) [reply memberbooth.py]: I agree

File: src/util/key_reader.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+            match = re.match(r"^DECODED: MANCHESTER=(0x[a-fA-F0-9]{8})$", line)
+            if match is not None:
+                complete_readouts.append(match.group(1))
+        if len(complete_readouts) == 0:
+            return False
+        self.last_tag_id = complete_readouts[-1]
+        return True
+
+    def get_aptus_tag_id(self):
+        aptus_tag_number = int(self.last_tag_id, 16) % 1000000000
```
Comment (InMyOrbit) [inline src/util/key_reader.py]: Based on our discussions, int(1e9) would increase readability :-).

File: src/util/key_reader.py (hunk 6 above)
Comment (InMyOrbit) [reply src/util/key_reader.py]: Save it for a rainy day.

File: src/util/key_reader.py (hunk 8 above)
Comment (emanuelen5) [reply src/util/key_reader.py]: This should be fixed as well

File: src/util/key_reader.py (hunk 6 above)
Comment (emanuelen5) [reply src/util/key_reader.py]: Fixed

File: src/util/key_reader.py (hunk 3 above)
Comment (emanuelen5) [reply src/util/key_reader.py]: Replacing magic constant with actual number of bytes in pipe

File: src/util/key_reader.py (hunk 8 above)
Comment (emanuelen5) [reply src/util/key_reader.py]: Fixed

File: memberbooth.py (hunk 7 above)
Comment (emanuelen5) [reply memberbooth.py]: I think we should add back the regular keyboard input (that worked with the Aptus keyboard). However, it is better if this is a separate flag (like `--tag-reader-type=APTUS`, `--tag-reader-type=MANUAL` and `--tag-reader-type=EM4100`). The `--development` flag is stated to be for events.

**Comment by emanuelen5** [comment]
I fixed most of the review comments.

File: memberbooth.py (hunk 7 above)
Comment (InMyOrbit) [reply memberbooth.py]: Could you add that as an issue and add it to the --development flag for now? I think the --tag-reader flag requires a bigger update to the states.py file.

Comment (emanuelen5) [reply memberbooth.py]: > If --development is used, I think that the application should run even if there is no reader attached.

But what should it do? Just mock the whole reader and not allow input? Then that could be a separate flag `--no-reader`? (Like the other mock flags)

**Pull request merged**